core/solver/gmres.hpp

```diff
@@ -91,7 +88,7 @@ class Gmres : public EnableLinOp<Gmres<ValueType>>,
      *
      * @return the preconditioner operator used by the solver
      */
-    std::shared_ptr<const LinOp> get_preconditioner() const
+    std::shared_ptr<const LinOp> get_preconditioner() const override
     {
         return preconditioner_;
```

Comment:
Question: does having a `preconditioner_` then become a property of being "preconditionable" or do we keep this as a property of the solver. Both could make sense I guess.

Reply:
This is a good question. We do something similar for LinOp, where the size is a part of the base (and I'm also not 100% sure we made the right decision there).

The upside of `preconditioner_`  being a member of the interface is that `get_preconditioner()` no longer has to be virtual, so the compiler can inline the call instead of going through a double redirection via the virtual table.

The downside is that the object in question needs to store the preconditioner in a variable called `preconditioner_`. That may not always be the best idea. For example, maybe it's better to compute the preconditioner on the fly - this design doesn't allow that. Maybe the solver already has the preconditioner stored as a different variable - then we have to duplicate this variable.

Reply:
If they compute a preconditioner on the fly, they probably would not use the `Preconditionable` interface because the solver is implicitly preconditioned directly in the code (in apply's code for example), no?

Reply:
That does not mean you cannot compute the preconditioner explicitly if the user requests it...

Reply:
So yes, if it's preconditioned, even implicitly, it may still be useful to provide a way of obtaining the preconditioner.

Reply:
Then you could still overload the `get_preconditioner` function and set the `preconditioner_` parameter to be empty (I believe it's a smart pointer in which case `nullptr` would work).

Reply:
I don't get it...

1. We don't want the preconditioner to be empty.
2. Even if we did: You probably mean "override" and not "overload"? I'm not sure how overloading anything would help. The problem with the "override" solution is that it would have the downsides of both approaches - the `get_preconditioner()` function would have to be virtual, which means no compiler optimizations are possible, which was the only upside of doing it in the first place (unless I'm missing something).

The case for the performance improvement is something like:

```c++
auto p = dynamic_cast<Preconditionable *>(a_lin_op);
for (auto i = 0u; i < a_large_number; ++i) {
    do_something_with(p->get_preconditioner()); // a lot of function calls
}
```

which is actually just a safeguard for user stupidity, since the same can be written more efficiently as:

```c++
auto p = dynamic_cast<Preconditionable *>(a_lin_op);
auto prec = p->get_precondiitoner();  // negligible - far less expensive than dynamic_cast
for (auto i = 0u; i < a_large_number; ++i) {
    do_something_with(prec);
}
```

If we set aside performance implications for the moment, and look at it purely from the software design point of view, I would strongly argue for the option we have implemented now:

The reason is that this option presents an interface that says:
> Preconditionable is an object that has a preconditioner, obtainable through the `get_preconditioner()` method.

The other option reads:
> Preconditionable is an object that has a preconditioner, stored in a `preconditionable_` protected data member, and is obtainable through the `get_preconditioner()` method, which returns the value stored in `preconditionable_`.

The second one has an obvious problem: `Preconditionable` is an interface, thus it defines how the object communicates with the outside world. Why would that need to have anything to do with the internals of an object (it's data members and implementations of its methods)?
In many other languages (usually the ones without multiple inheritance), you wouldn't even be able to do option 2, as an interface would be limited to only method declarations.

So, the only 2 possibilities I see are the following:

```c++
// proper interface for the software design, may lose some performance
// if the users are not aware about performance cost of virtual methods
class Preconditionable() {
public:
    virtual ~Preconditionable() = default;
    virtual shared_ptr<const LinOp> get_preconditioner() const = 0;
};
```

```c++
// workaround for performance in corner cases,
// bad from the software design point of view
class Preconditionable() {
public:
    virtual ~Preconditionable() = default;
    const shared_ptr<const LinOp>& get_preconditioner() const { return preconditioner_; }
protected:
    shared_ptr<const LinOp> preconditioner_;
};
```

The question is how much flexibility and good software design practices are we willing to sacrifice to get more performance from suboptimal user code.

Reply:
Option 2 is moving away from providing an interface towards providing a reusable base class and object which encapsulates standard methods and members for adding and manipulating a preconditioner to an object/LinOp. This is a similar design to the `EnableLogger`. That was my question since the beginning anyway and I only wanted to bring that issue for @all.

Reply:
Yes going to the direction of using a mixin was more what I was hinting at, also because we are using mixins for pretty much everything at this point so that would keep some form of uniformity. I understand that with the current code and the fact that the impact of the `Preconditionable` class is isolated in 5 solvers there may be no need to move too much towards that direction yet.